### PR TITLE
CORE-8602: x264 on robot

### PR DESCRIPTION
**Pull request by ghost**
builds with catkin using the apt-get version of ffmpeg
fixed or silenced errors that spam the console/log

### CODE REVIEW

File: src/x264_subscriber.cpp [hunk 1]
```diff
@@ -92,18 +92,20 @@ namespace x264_image_transport {
...
 
 		// Assign appropriate parts of buffer to image planes in pFrameRGB
 		// Note that pFrameRGB is an AVFrame, but AVFrame is a superset
 		// of AVPicture
-		avpicture_fill((AVPicture *)codecs[i].m_pFrameRGB, codecs[i].m_buffer, PIX_FMT_RGB24,
+		avpicture_fill((AVPicture *)codecs[i].m_pFrameRGB, codecs[i].m_buffer, AV_PIX_FMT_RGB24,
 		               codecs[i].m_pCodecCtx->width, codecs[i].m_pCodecCtx->height);
 		codecs[i].m_img_convert_ctx = NULL;
+
+        av_log_set_level(AV_LOG_QUIET); // silence all the errors/warnings
```
Comment (servos) [inline src/x264_subscriber.cpp]: spacing here looks wrong. I think you need to use tabs. Since the rest of the file uses tabs just match that. Normally in our code we use spaces and never tabs.

**Review by servos — APPROVED** [review]

File: FindFFMPEG.cmake [hunk 2]
```diff
@@ -32,32 +32,32 @@ else (FFMPEG_LIBRARIES AND FFMPEG_INCLUDE_DIR)
 
   find_path(FFMPEG_AVCODEC_INCLUDE_DIR
     NAMES libavcodec/avcodec.h
-    PATHS $ENV{HOME}/ffmpeg_build/include ${_FFMPEG_AVCODEC_INCLUDE_DIRS} /usr/include /usr/local/include /opt/local/include /sw/include
```
Comment (efernandez) [inline FindFFMPEG.cmake:35]: Should be good, but maybe it's worth trying this other .cmake file instead of this one: https://github.com/snikulov/cmake-modules/blob/master/FindFFmpeg.cmake (optional)

File: src/x264_publisher.cpp [hunk 3]
```diff
@@ -241,48 +248,49 @@ namespace x264_image_transport {
...
         //Initialize packet
         av_init_packet(&encodedPacket_);
         encodedPacket_.data = NULL;    // packet data will be allocated by the encoder
         encodedPacket_.size = 0;
-			
+
 		initialized_ = true;
 		ROS_INFO("x264Publisher::initialize_codec(): codec initialized (width: %i, height: %i, fps: %i)",width,height,fps);
         pthread_mutex_unlock (&mutex_);
+
```
Comment (efernandez) [inline src/x264_publisher.cpp]: Fix indentation (also for the instruction above although it's not part of the PR)

File: FindFFMPEG.cmake (hunk 2 above)
Comment (efernandez) [reply FindFFMPEG.cmake:35]: You can try this after this PR, if you want

File: src/x264_publisher.cpp [hunk 4]
```diff
@@ -192,37 +192,44 @@ namespace x264_image_transport {
...
             pthread_mutex_unlock (&mutex_);
             return;
         }
 
+        encFrame_->width = width;
+        encFrame_->height = height;
+
         // Prepare the software scale context
         // Will convert from RGB24 to YUV420P
         if (encoding == enc::BGR8)
```
Comment (efernandez) [inline src/x264_publisher.cpp]: Not for this PR, but it seems that this could be simplified a lot doing this:
``` c++
if (encoding == enc::BGR8)
{
  encFrame_->format = AV_PIX_FMT_BGR24;
} else if (encoding == enc::RGB8)
{
  encFrame_->format = AV_PIX_FMT_RGB24;
}  ...
```

for all "good" branches, and then:

``` c++
            sws_ctx_ = sws_getContext(width, height, encFrame_->format, //src
                                      encCdcCtx_->width, encCdcCtx_->height, encCdcCtx_->pix_fmt, //dest
                                      SWS_FAST_BILINEAR, NULL, NULL, NULL);
```

Of even creating an std::map with all the mappings, so instead of the if-else-if blocks, you can just do something like this:
``` c++
const auto format = encodingFormatsMap.find(encoding);

if (format == encodingFormatsMaps.end())
{
  if (encoding == enc::BAYER_GBRG16)
  {
    // THE CODE FOR THIS NOT SUPPORTED YET ENCODING
  }
  else
  {
    // THE CODE FOR UNKNOWN NON-SUPPORTED ENCODING
  }
}
```

**Review by efernandez — APPROVED** [review]
Only minor comments and some suggestions for future PRs (optionally).

LGTM

File: cfg/x264Publisher.cfg [hunk 5]
```diff
@@ -6,6 +6,6 @@ from dynamic_reconfigure.parameter_generator_catkin import *
 
 gen = ParameterGenerator()
 
-gen.add("qmax", int_t, 0, "Maximum quantizer, less is better image quality",51,0,63)
+gen.add("qmax", int_t, 0, "Maximum quantizer, less is better image quality",30,0,63)
```
Comment (efernandez) [inline cfg/x264Publisher.cfg:9]: You can use `rosrun rqt_reconfigure rqt_reconfigure` and play with the slider to try it live.
